### PR TITLE
ref(hc): Guard transaction.atomic and get_connection

### PR DESCRIPTION
**Pull request by corps**
Adds flake8 support for transaction.atomic and get_connection not specifying `using`.

Todo: run this on the sentry and getsentry codebases, fixing existing issues.
Notably, it is technically safe to commit this before all existing are fixed, as our flake8 only run on changed files.  Still, I can spend some time on getting existing ones swapped over.

### CODE REVIEW

File: tools/flake8_plugin.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 from __future__ import annotations
 
 import ast
-from typing import Any, Generator
+from typing import Any, Generator, List
```
Comment (asottile-sentry) [inline tools/flake8_plugin.py:4]: don't use `List`, use `list`

File: tools/flake8_plugin.py [hunk 2]
```diff
@@ -61,6 +82,28 @@ def visit_Name(self, node: ast.Name) -> None:
...
         self.generic_visit(node)
 
+    def visit_Call(self, node: ast.Call) -> None:
+        name_list: List[str] = []
+        func = node.func
+        while isinstance(func, ast.Attribute):
+            name_list.append(func.attr)
+            func = func.value
+        if not isinstance(func, ast.Name):
+            return
```
Comment (asottile-sentry) [inline tools/flake8_plugin.py:92]: this is broken, it will not visit the members inside

Comment (asottile-sentry) [inline tools/flake8_plugin.py:92]: this seems over complicated -- this should probably inspect the expected attributes exactly?

File: tools/flake8_plugin.py [hunk 3]
```diff
@@ -38,13 +48,24 @@ def visit_ImportFrom(self, node: ast.ImportFrom) -> None:
             ):
                 self.errors.append((node.lineno, node.col_offset, S005_msg))
 
+            for name in node.names:
+                self.names_stack[-1][name.asname or name.name] = f"{node.module}.{name.name}"
```
Comment (asottile-sentry) [inline tools/flake8_plugin.py:52]: rather than this -- which is error prone (can be reassigned, etc.) let's just ban `as` with these names and then check the names directly below

it's never going to be perfect and it's currently very hard to follow what this is attempting to do

**Review by asottile-sentry — CHANGES_REQUESTED** [review]
this needs tests

**Comment by asottile-sentry** [comment]
also the currently failing things should be fixed before this is merged -- otherwise this will create significant disruption for developers.  the error message also isn't super clear what to fix so it'll be quite annoying out of context to try and resolve it

File: tools/flake8_plugin.py (hunk 1 above)
Comment (asottile-sentry) [reply tools/flake8_plugin.py:4]: we can, but in some parts of the codebase it breaks things -- notably anywhere `pydantic` has proliferated -- here it does not

File: tools/flake8_plugin.py (hunk 3 above)
Comment (corps) [reply tools/flake8_plugin.py:52]: The motivating issue is that `get_connection` is a common name; it's already used by `sentry.utils.email.send`.  Would it be preferable to rename that existing utility + ban naming functions with colliding names?

Comment (asottile-sentry) [reply tools/flake8_plugin.py:52]: yeah that sounds like a good way forward

Comment (corps) [reply tools/flake8_plugin.py:52]: Also, I do worry about other libraries that happen to have `get_connection` named functions and disambiguating them.  Using a module lookup here ensures we don't mix such other functions when they are used.

Any thoughts?

Comment (markstory) [reply tools/flake8_plugin.py:52]: How would we handle `get_connection` being defined in a library where we can't rename it though? Matching/resolving the module path through local renames allows us to stay compatible with more use cases.

Comment (asottile-sentry) [reply tools/flake8_plugin.py:52]: 1. make a `sentry_get_connection` for the intended function
2. ban aliasing of that
3. ban importing of the original name
4. ignore `get_connection` imports from unrelated modules

Comment (markstory) [reply tools/flake8_plugin.py:52]: 🤔 That seems heavy handed for what we're trying to do with the flake8 warning. We'll have 'hard' enforcement of the need for `using` done with runtime code. We have been thinking of the lint rule as a good way to give developers faster feedback as lint checks can happen locally where as CI results take a while.

Comment (asottile-sentry) [reply tools/flake8_plugin.py:52]: let's lean on the type checker instead -- this patch is trying to turn flake8 into one which it is not designed to do (and why this patch is so complex / hard to follow / doesn't catch all the cases)

Comment (corps) [reply tools/flake8_plugin.py:52]: The goal is not to turn flake8 into a type checker.  It's a concession to the reality of the names involved and their common collisions.  Preferably we avoid false positives, even if it means false negatives -- it's best effort.  I'm ok missing some cases if it catches the vast majority of cases, and also doesn't restrict too much what can be done by developers when dealing with name collisions (even `transaction` is a common name, so the way that the helpers are imported varies by module).

I'd love to lean into the type check if possible, but that doesn't seem possible.
1.  Most of our codebase still doesn't mypy check still.  If our codebase had full typing coverage, I think we could remove the lint rule.
2.  Is it possible to override the interface typing already providing by django?  A key issue is that the existing interfaces already allow for the `using` as an optionality -- I'm not aware if it is possible to force an "override" of the signature provided by django typings.

As for your concerns about complexity, I can refactor to readibility and improve tests, I don't mind.  But I still don't see a viable alternative.

Comment (asottile-sentry) [reply tools/flake8_plugin.py:52]: > I'm ok missing some cases if it catches the vast majority of cases

flake8 is not the right tool for what you're trying to do

> Most of our codebase still doesn't mypy check still

this isn't true, the majority does type check

> Is it possible to override the interface typing already providing by django

this is why I'm recommending we provide a different function entirely such that we have our own interface we can control and enforce

Comment (markstory) [reply tools/flake8_plugin.py:52]: > this is why I'm recommending we provide a different function entirely such that we have our own interface we can control and enforce

I agree that an application wrapper for transactions would make enforcing the `using` parameter to be present simpler, but it adds cognitive load for all future work as folks have to remember to use the application wrapper. Having to remember to provide `using` for all transactions is also additional cognitive load, but my thinking was that adding constraints to the existing functions was less overhead than having folks need to learn & remember an entirely new sentry-specific extension.

Comment (asottile-sentry) [reply tools/flake8_plugin.py:52]: some other ideas from the type checker approach -- assuming we keep the same name -- there's two ~easy-ish options:
- fork `django-stubs` (I've already been considering this due to [this](https://github.com/typeddjango/django-stubs/blob/cef666da8a155ad4bd380fa9157429d277c64237/django-stubs/db/models/fields/__init__.pyi#L189-L191) playing poorly with functions taking `HC | Model` instances
    - tl;dr would be to setup a `sentry-django-stubs` fork, publish that to public and internal pypi via `craft`, replace out the dependency in sentry, synchronize the fork over time
    - this has the downside in that we'd have to maintain the patches on top of `django-stubs` indefinitely -- the ones I had in mind would hopefully be fixed by mypy fixing union resolutions
- write a mypy plugin for the particular attributes -- this actually wouldn't be too bad for this ~relatively simple case but mypy plugins can get gnarly quickly.   you could also potentially use this to do gradual rollout via our existing gradual mypy machinery by introducing a special error code for your particular violation

Comment (markstory) [reply tools/flake8_plugin.py:52]: If we're going to have a fork of django-stubs, we can piggyback the additional `transaction.atomic()` constraints there. Having to maintain that going forwards isn't great though. I can look into a mypy plugin and see how that fares.
